client/network/src/service.rs

```diff
@@ -571,12 +587,50 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 	/// parameter is a `&'static str`, and not a `String`, in order to avoid accidentally having
 	/// an unbounded set of Prometheus metrics, which would be quite bad in terms of memory
 	pub fn event_stream(&self, name: &'static str) -> impl Stream<Item = Event> {
-		// Note: when transitioning to stable futures, remove the `Error` entirely
 		let (tx, rx) = out_events::channel(name);
 		let _ = self.to_worker.unbounded_send(ServiceToWorkerMsg::EventStream(tx));
 		rx
 	}
 
+	/// Sends a single targeted request to a specific peer. On success, returns the response of
+	/// the peer.
+	///
+	/// Request-response protocols are a way to complement notifications protocols, but
+	/// notifications should remain the default ways of communicating information. For example, a
+	/// peer can announce something through a notification, after which the recipient can obtain
+	/// more information by performing a request.
+	/// As such, this function is meant be called only with peers we are already connected to.
+	/// Calling this method with a `target` we are not connected to will *not* attempt to connect
+	/// to said peer.
+	///
+	/// Contrary to notifications, requests are guaranteed to not be interrupted under normal
+	/// circumstances.
```

Comment:
Not quite sure I understand this comment. How are notifications not guaranteed to not be interrupted under normal circumstances?

Reply:
What I wanted to say with this comment is "in the future, the peerset might decide to close connections and discard pending notifications, while request-responses won't suffer from that" but without exposing these details.

Reply:
I would suggest either exposing these details or removing the comment as I find the comment without the details more confusing then helpful.

Reply:
> But if a connection is closed for whatever reason while there are pending requests (that is, requests that were sent while the connection was still established and are waiting for responses), these will be aborted. So I'm also not sure what the distinction is supposed to be here?

The terminology is very confusing here.
According to the peerset, "connected" doesn't mean "has a TCP connection" but "has a gossiping substream".
Similarly, when the peerset "disconnects" a peer, it actually means "close the gossiping substream", and not necessarily the connection.

But to add to the confusion, the legacy substream closes the entire connection when it closes, as a way to avoid a potential race condition when reopening a substream. Consequently, "closing the substream", at the moment, also means "closing the connection". This will no longer be the case in the future.

Reply:
To go back to the comment in the PR: what I want to express is that closing the gossiping substream doesn't interrupt ongoing requests.
If you simulate request-responses on top of notifications, a decision from the peerset to close the gossiping substream would interrupt these requests.

Reply:
Thanks, that clarifies it for me - not sure how best to express that in the code comment. Just one further clarification: When you refer to "the gossiping substream" and e.g. "closing the gossiping substream" is that a specific notifications protocol substream or does this terminology refer to all of them?

Reply:
> When you refer to "the gossiping substream" and e.g. "closing the gossiping substream" is that a specific notifications protocol substream or does this terminology refer to all of them? (My understanding is now that, at the moment, it also refers to the "legacy substream", which still closes the connection when it gets closed, but that won't be the case in the future since it will be removed).

At the moment this indeed refers to all of them.

In the future, the intention is that, depending on the peer, potentially only a subset of the notifications substreams would be open.
For example, a validator connected to a validator would open an additional notifications protocol to communicate parachain attestation messages, which is something that is only relevant to validators.

Mentioning "the" gossiping substream would therefore be misleading in the future.

Reply:
I've opted to drop the PR comment entirely.